### PR TITLE
[SPARK-23702][SS] Forbid watermarks on both sides of stateful streaming operators.

### PR DESCRIPTION
**Pull request by jose-torres**
## What changes were proposed in this pull request?

Forbid watermarks on both sides of stateful streaming operators.

Multiple sequential watermarks are in general not supported by the execution engine; support is only in parallel, e.g. on both sides of a join. We can normally resolve this by simply picking the topmost watermark operator  and ignoring the rest, but this is not semantically valid when there's a stateful operator in between.

## How was this patch tested?

new unit test

### CODE REVIEW

**Comment by jose-torres** [comment]
@tdas

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 1]
```diff
@@ -160,6 +160,19 @@ object UnsupportedOperationChecker {
         case _: InsertIntoDir =>
           throwError("InsertIntoDir is not supported with streaming DataFrames/Datasets")
 
+        case e: EventTimeWatermark =>
+          val statefulChildren = e.collect {
+            case a: Aggregate if a.isStreaming => a
+            case d: Deduplicate if d.isStreaming => d
+            case f: FlatMapGroupsWithState if f.isStreaming => f
```
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: Should be for joins as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala [hunk 2]
```diff
@@ -160,6 +160,19 @@ object UnsupportedOperationChecker {
...
+          }
+          statefulChildren.foreach { statefulNode =>
+            if (statefulNode.collectFirst{ case e: EventTimeWatermark => e }.isDefined) {
+              throwError("Watermarks both before and after a stateful operator in a streaming " +
```
Comment (tdas) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: This gives the impression that it makes sense but we dont support it. In fact, its just ill-defined. May change this to something like ... Multiple watermarks before and after stateful operators is not well-defined in a streaming query.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala [hunk 3]
```diff
@@ -140,6 +140,21 @@ class UnsupportedOperationsSuite extends SparkFunSuite {
...
+      CalendarInterval.fromString("interval 1 second"),
+      Aggregate(
+        attributeWithWatermark :: Nil,
+        aggExprs("a"),
+        EventTimeWatermark(
+          attribute,
+          CalendarInterval.fromString("interval 2 seconds"),
+          streamRelation))),
+    outputMode = Append,
+    expectedMsgs = Seq("both before and after"))
```
Comment (tdas) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationsSuite.scala]: Add for other stateful operators as well.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala (hunk 2 above)
Comment (jose-torres) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/UnsupportedOperationChecker.scala]: WDYT of something like "watermarks may not be present...". Talking about "well-defined" seems a bit confusing to me.

**Comment by HeartSaVioR** [comment]
How would we like to handle this patch? I guess we add feature on handling multiple watermarks in #21701 so based on the direction this patch might be going to be abandoned. IMHO I'm not 100% sure we have clear use cases for defining multiple watermarks for one source.

**Pull request closed without merging**